### PR TITLE
Replaces `NotImplementedException` with `NotSupportedException`

### LINKED ISSUE #292: [Proposal] Replace all NotImplementedException throws with NotSupportedException
# [CHANGE] Replace all `NotImplementedException` throws with `NotSupportedException`

* [x] Proposed
* [x] Prototype
* [ ] Implementation: Not Started
  * [ ] iOS Support
  * [ ] Android Support
  * [ ] macOS Support
  * [ ] Windows Support
* [ ] Unit Tests: Not Started
* [ ] Sample: Not Started
* [ ] Documentation: Not Started

## Link to Discussion

Came out of PR #289

## Summary
[summary]: #summary

Replace all `NotImplementedException` throws with `NotSupportedException` to give a clearer indication that something is not supported rather than not supported yet or we forgot to implement it.

## Motivation
[motivation]: #motivation

To provide a clearer indication to consumers that we have decided not to support something.

## Detailed Design
[design]: #detailed-design

~`throw new NotImplementedException();`~

becomes

`throw new NotSupportedException("Meaningful reason why.");`

**Bonus points:** find a way to make usage of `NotImplementedException` a warning/error. Likely a `.editorconfig` rule if possible.

## Usage Syntax
[usage]: #usage-syntax

N/A

## Drawbacks
[drawbacks]: #drawbacks

None I can think of.

## Alternatives
[alternatives]: #alternatives

Do nothing?

## Unresolved Questions
[unresolved]: #unresolved-questions

None I can think of.

### PR DESCRIPTION
**Pull request by bijington**
### Description of Change ###

Replaces all usages of `NotImplementedException` with `NotSupportedException`.
 

 ### Linked Issues ###
 

 - Fixes #292

 ### PR Checklist ###
 
 - [ ] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [ ] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls/XYZ 

 ### Additional information ###

### CODE REVIEW

**Comment by bijington** [comment]
Some comments online suggest we may be able to prevent the usage of `NotImplementedException` (https://stackoverflow.com/a/69689012/32348) however they do point to [CA1031](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1031) which looks aimed at preventing its use inside a `catch` rather than specifically preventing the throwing of it.

**Comment by bijington** [comment]
The introduction of `BaseOneWayConverter` may or may not complicate the support of `null`s inside converters as mentioned here: https://github.com/CommunityToolkit/Maui/discussions/164. Note while this discussion is closed the functionality was added to the Xamarin Community Toolkit.

**Comment by brminnick** [comment]
Hey Shaun! I just had a moment of inspiration about adding `null` support so I figured I'd open up a PR to help explain what I'm thinking. 

Check it out and let me know what you think! 

https://github.com/CommunityToolkit/Maui/pull/310/files

**Comment by brminnick** [comment]
> Some comments online suggest we may be able to prevent the usage of `NotImplementedException` (https://stackoverflow.com/a/69689012/32348) however they do point to [CA1031](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1031) which looks aimed at preventing its use inside a `catch` rather than specifically preventing the throwing of it.

Dang - yea, I can't seem to find anything either that would cause the compiler to throw an error when it detects `NotImplementedException`. 

Since this PR is growing a bit in scope, I recommend doing the following, then let's merge this PR:
- Update the Contribution Guidelines to recommend against `NotImplementedException`
- Open a new `[Proposal]` that focuses on preventing `NotImplementedException` via `editorconfig`
  - We'll leave this open as an unknown
  - Maybe someday in the future we, or someone smarter than us, can figure out how to do it

**Comment by brminnick** [comment]
> - Open a new Discussion that focuses on preventing `NotImplementedException` via `editorconfig`
>  - Maybe someday in the future we, or someone smarter than us, can figure out how to do it

Discussion opened: https://github.com/CommunityToolkit/Maui/discussions/313

**Comment by brminnick** [comment]
Hey @bijington! I've updated `Contributing.md` to include `NotImplementedException`.

I'm going to move this PR out of Draft because I think it's ready to be merged now that we've moved the `NotImplementedException` discussion to #313.

**Review by brminnick — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -182,6 +182,12 @@ if (something is not null)
 }
 ```
 
+### `NotImplementedException`
+
+Please avoid adding new code that throws a `NotImplementedException`. According to the [Microsoft Docs](https://docs.microsoft.com/dotnet/api/system.notimplementedexception?WT.mc_id=mobile-0000-bramin), we should only "throw a `NotImplementedException` exception in properties or methods in your own types when the that member is still in development and will only later be implemented in production code. In other words, a NotImplementedException exception should be synonymous with 'still in development.'"
```
Comment (bijington) [inline CONTRIBUTING.md:187]:
```suggestion
Please avoid adding new code that throws a `NotImplementedException`. According to the [Microsoft Docs](https://docs.microsoft.com/dotnet/api/system.notimplementedexception), we should only "throw a `NotImplementedException` exception in properties or methods in your own types when that member is still in development and will only later be implemented in production code. In other words, a NotImplementedException exception should be synonymous with 'still in development.'"
```

**Review by bijington — COMMENTED** [review]
@brminnick thanks for the updates to this PR :)

**Review by brminnick — APPROVED** [review]
Thanks Shaun! I'm good for you to merge this 🙌

**Pull request merged**